src/target/riscv/riscv-013.c

```diff
@@ -3396,329 +3396,567 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
 
-	int result = ERROR_OK;
+	if (read_memory_progbuf_inner_fill_progbuff(target, access.increment,
+				access.element_size, mprven) != ERROR_OK)
+		return ERROR_FAIL;
 
-	/* Write address to S0. */
-	result = register_write_direct(target, GDB_REGNO_S0, address);
-	if (result != ERROR_OK)
-		return result;
+	if (read_memory_progbuf_inner_startup(target, access.target_address,
+				access.increment, /*index*/ 0)
+			!= ERROR_OK)
+		return ERROR_FAIL;
+	/* The program in program buffer is executed twice during
+	 * read_memory_progbuf_inner_startup().
+	 * Here:
+	 * dmi_data[0:1] == M[address]
+	 * s1 == M[address + increment]
+	 * s0 == address + increment * 2
+	 * `count - 2` program executions are performed in this loop.
+	 * No need to execute the program any more, since S1 will already contain
+	 * M[address + increment * (count - 1)] and we can read it directly.
+	 */
+	const uint32_t loop_count = count - 2;
+
+	for (uint32_t index = 0; index < loop_count;)
+		if (read_memory_progbuf_inner_loop_iteration(target, access, &index,
+					loop_count) != ERROR_OK) {
+			dmi_write(target, DM_ABSTRACTAUTO, 0);
+			return ERROR_FAIL;
+		}
+	if (dmi_write(target, DM_ABSTRACTAUTO, 0) != ERROR_OK)
+		return ERROR_FAIL;
 
-	if (increment == 0 &&
-			register_write_direct(target, GDB_REGNO_A0, 0) != ERROR_OK)
+	/* Read the penultimate word. */
+	if (read_word_from_dmi(target, access, count - 2)
+			!= ERROR_OK)
+		return ERROR_FAIL;
+	/* Read the last word. */
+	return read_word_from_s1(target, access, count - 1);
+}
+/**
+ * Only need to save/restore one GPR to read a single word, and the progbuf
+ * program doesn't need to increment.
+ */
+static int read_memory_progbuf_inner_one(struct target *target,
+		struct memory_access_info access, bool mprven)
+{
+	if (riscv_save_register(target, GDB_REGNO_S1) != ERROR_OK)
 		return ERROR_FAIL;
 
+	struct riscv_program program;
+
+	riscv_program_init(&program, target);
+	if (riscv_program_load_mprv(&program, GDB_REGNO_S1, GDB_REGNO_S1, 0,
+				access.element_size, mprven) != ERROR_OK)
+		return ERROR_FAIL;
+	if (riscv_program_ebreak(&program) != ERROR_OK)
+		return ERROR_FAIL;
+	if (riscv_program_write(&program) != ERROR_OK)
+		return ERROR_FAIL;
+
+	/* Write address to S1, and execute buffer. */
+	if (write_abstract_arg(target, 0, access.target_address, riscv_xlen(target))
+			!= ERROR_OK)
+		return ERROR_FAIL;
 	uint32_t command = access_register_command(target, GDB_REGNO_S1,
-			riscv_xlen(target),
+			riscv_xlen(target), AC_ACCESS_REGISTER_WRITE |
 			AC_ACCESS_REGISTER_TRANSFER | AC_ACCESS_REGISTER_POSTEXEC);
 	if (execute_abstract_command(target, command) != ERROR_OK)
 		return ERROR_FAIL;
 
-	/* First read has just triggered. Result is in s1. */
-	if (count == 1) {
-		uint64_t value;
-		if (register_read_direct(target, &value, GDB_REGNO_S1) != ERROR_OK)
+	return read_word_from_s1(target, access, 0);
+}
+
+static void set_buffer_and_log_access(struct memory_access_info access,
```

Comment:
```suggestion
static void set_buffer_and_log_read(struct memory_access_info access,
```

Reply:
I would like to leave this function's name as is.

The only thing this function does is calling `buf_set_u64` and `log_memory_access64` with corresponding arguments. So I would like for the name to represent this.

Reply:
Yes, but it only logs a memory read. It is not used for arbitrary memory accesses.